include/envoy/stats/stats.h

```diff
@@ -408,25 +408,33 @@ typedef std::unique_ptr<StoreRoot> StoreRootPtr;
 struct RawStatData;
 
 /**
- * Abstract interface for allocating a RawStatData.
+ * Abstract interface for allocating statistics. Alternate
+ * implementations can be created utilizing a single fixed-size block
+ * suitable for shared-memory, or in the heap, allowing for pointers
+ * and sharing of substrings, with an opportunity for reduced memory
+ * consumption.
  */
-class RawStatDataAllocator {
+class StatDataAllocator {
 public:
-  virtual ~RawStatDataAllocator() {}
+  virtual ~StatDataAllocator() {}
 
   /**
-   * @return RawStatData* a raw stat data block for a given stat name or nullptr if there is no
-   *         more memory available for stats. The allocator should return a reference counted
-   *         data location by name if one already exists with the same name. This is used for
-   *         intra-process scope swapping as well as inter-process hot restart.
+   * @param name the full name of the stat.
+   * @param tag_extracted_name the name of the stat with tag-values stripped out.
+   * @param tags the extracted tag values.
+   * @return Counter* a counter, or nullptr if allocation failed.
    */
-  virtual RawStatData* alloc(const std::string& name) PURE;
+  virtual Counter* makeCounter(const std::string& name, std::string& tag_extracted_name,
```

Comment:
It seems that we removed the `free()`, but we're still returning raw pointers.  Can you document the ownership model?  How is the `StatDataAllocator` informed of when the stat is no longer needed by the caller?  Is the user allowed to just delete the raw pointer?

Reply:
Very good point here, and I think my design here needs a bit of rethinking in the context of how a caller would free stuff.  The other comments I'll address later, and I'm flipping back to WIP-mode while I think about this :)

Reply:
Actually this was easy to deal with by just returning a shared_ptr at this level.